text/0000-api-evolution.md

```diff
@@ -0,0 +1,748 @@
+items trump any trait items with the same name.
+
+However, introducing an inherent item *can* lead to breakage if the signature of
+the item does not match that of an in scope, implemented trait:
+
+```rust
+// Crate A
+    pub struct Foo;
+
+// Crate B
+    trait Trait {
+        fn foo(&self);
+    }
+
+    impl Trait for crateA::Foo { .. }
+
+    fn use_foo(f: &crateA::Foo) {
+        f.foo()
+   }
+```
+
+If crate A adds a method:
+
+```rust
+impl Foo {
+    fn foo(&self, x: u8) { ... }
+}
+```
+
+then crate B would no longer compile, since dispatch would prefer the inherent
+impl, which has the wrong type.
+
+Once more, this is considered a minor change, since UFCS can disambiguate (see
+"Adding a defaulted item" above).
+
+It's worth noting, however, that if the signatures *did* happen to match then
+the change would no longer cause a compilation error, but might silently change
+runtime behavior. The case where the same method for the same type has
+meaningfully different behavior is considered unlikely enough that the RFC is
+willing to permit it to be labeled as a minor change -- and otherwise, inherent
+methods could never be added after the fact.
+
+### Other items
+
+Most remaining items do not have any particularly unique items:
+
+* For type aliases, see "[Signatures in type definitions](#signatures-in-type-definitions)".
+* For free functions, see "[Signatures in functions](#signatures-in-functions)".
+
+## Cross-cutting concerns
+
+### Behavioral changes
+
+This RFC is largely focused on API changes which may, in particular, cause
+downstream code to stop compiling. But in some sense it is even more pernicious
+to make a change that allows downstream code to continue compiling, but causes
+its runtime behavior to break.
+
+This RFC does not attempt to provide a comprehensive policy on behavioral
+changes, which would be extremely difficult. In general, APIs are expected to
+provide explicit contracts for their behavior via documentation, and behavior
+that is not part of this contract is permitted to change in minor
+revisions. (Remember: this RFC is about setting a *minimum* bar for when major
+version bumps are required.)
+
+This policy will likely require some revision over time, to become more explicit
+and perhaps lay out some best practices.
+
+### Signatures in type definitions
+
+#### Major change: tightening bounds.
+
+Adding new constraints on existing type parameters is a breaking change, since
+existing uses of the type definition can break. So the following is a major
+change:
+
+```rust
+// MAJOR CHANGE
+
+// Before
+struct Foo<A> { .. }
+
+// After
+struct Foo<A: Clone> { .. }
+```
+
+#### Minor change: loosening bounds.
+
+Loosening bounds, on the other hand, cannot break code because when you
+reference `Foo<A>`, you *do not learn anything about the bounds on `A`*. (This
+is why you have to repeat any relevant bounds in `impl` blocks for `Foo`, for
```

Comment:
It's not completely clear, but this RFC can be revised if or when [implied bounds](http://smallcultfollowing.com/babysteps/blog/2014/07/06/implied-bounds/) happen.